oauth2_http/java/com/google/auth/oauth2/IdentityPoolSubjectTokenProvider.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.auth.oauth2;
+
+import com.google.api.client.json.GenericJson;
+import com.google.api.client.json.JsonObjectParser;
+import com.google.auth.oauth2.IdentityPoolCredentialSource.CredentialFormatType;
+import com.google.common.io.CharStreams;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Serializable;
+import java.nio.charset.StandardCharsets;
+import java.util.function.Supplier;
+
+/**
+ * Provider for retrieving subject tokens for {@Link IdentityPoolCredentials} to exchange for GCP
+ * access tokens.
+ */
+abstract class IdentityPoolSubjectTokenProvider implements Serializable {
+
+  /**
+   * Gets a subject token that can be exchanged for a GCP access token.
+   *
+   * @return a valid subject token.
+   * @throws IOException
+   */
+  abstract String getSubjectToken() throws IOException;
+
+  /**
+   * Gets the metrics header value that should be used for the sts request.
+   *
+   * @return the metrics header value.
+   */
+  abstract String getMetricsHeaderValue();
+
+  /**
+   * Gets the supplier used to retrieve the subject token.
+   *
+   * @return the supplier used for getSubjectToken.
+   */
+  abstract Supplier<String> getSupplier();
```

Comment:
Maybe similar to @lsirac comment: we can probably remove getSupplier() and use getToken() directly (it would call the supplier in the programmatic case);

That said, it might be better if customers provided a `SubjectTokenProvider` directly.  The (internal) API allows us to cheat and deal with `IOException` in a way we don't allow via `Supplier<String>`;  I see both problems of us eating `IOException` and turning them into RTEs as well as us catching RTEs and turning them into not-retryable `IOException` -- even though we can anticipate that there will be network, IPC or File errors present based on all of our own implementations.

So...my suggestion is to ditch "Supplier" and offer "Provider" to our customers.  Now everyone has the same API `String getSubjectToken() throws IOException`; they can implement Retryable on their exceptions.  You might need to find another way to support the metrics header value if you're worried about clients overriding it, but overall the code should be simpler.

Reply:
Made this change - I wanted to keep the supplier so we could have a functional interface for users so they could use a lambda instead of creating a whole class, but I agree that its creating more problems than its worth. Refactored everything to be public. I named the objects IdentityPoolSubjectTokenSupplier and AwsSecurityCredentialSupplier instead of provider just because I call the similar function/object supplier in other PR's for other languages (still in flight, not merged) but don't feel super strongly about that either way.